main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -923,4 +993,24 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
     }
   }
 
+  /**
+    * @param defn  def whose context we are in
+    * @param arity number of parameters required to execute recursion
+    */
+  private case class RecursionContext(defn: Symbol.DefnSym, arity: Int, applies: Int = 0) {
```

Comment:
I wonder if the following design would be more clear and faster:

```
sealed trait RecursionContext {...}
object NoContext extends RecursionContext {...}
case class Apply0(defn: Symbol.DefnSym) extends RecursionContext {...}
case class ApplyN(defn: Symbol.DefnSym, applies: Int) extends RecursionContext {...}
```

Reply:
Do you see how each class can then easily implemented the addApply and resetApply methods?

Reply:
I like this design a lot. It does add a slight weirdness in the `Env.resetApplies` method, since in order to avoid extra object creation, I ended up pattern matching instead of simply delegating to `resetApplies` as might be expected.

Reply:
Did you notice any performance impact?

Reply:
None outside the usual variance.